src/infra/core/src/ingest/fetch_service/core.rs

```diff
@@ -345,3 +395,104 @@ impl std::convert::From<EventTimeSourceError> for PollingIngestError {
+        let template = ["${{ env.", &var, " }}"].concat();
+        (var, template)
+    }
+
+    fn ds_env_var(key: &str, val: &str) -> Result<DatasetEnvVar, DatasetEnvVarEncryptionError> {
+        DatasetEnvVar::new(
+            key,
+            Utc::now(),
+            &DatasetEnvVarValue::Regular(val.to_string()),
+            &DatasetID::new_seeded_ed25519(b"foo"),
+            SAMPLE_DATASET_ENV_VAR_ENCRYPTION_KEY,
+        )
+    }
+
+    #[test]
+    fn test_fetch_service_template_string() {
+        let temp_dir = tempfile::tempdir().unwrap();
+
+        let fetch_svc = FetchService::new(
+            Arc::new(ContainerRuntime::default()),
+            None,
+            None,
+            None,
+            None,
+            Arc::new(DatasetKeyValueServiceSysEnv::new()),
+            Arc::new(RunInfoDir::new(temp_dir.path().join("run"))),
+        );
+
+        let (var, template) = var_and_template();
+
+        let env_vars_empty: HashMap<String, DatasetEnvVar> = HashMap::new();
+        let env_vars: HashMap<String, DatasetEnvVar> =
+            HashMap::from([(var.clone(), ds_env_var(&var, "222").unwrap())]);
+
+        // When dataset_env_vars is empty and system env var is set,
+        // its value is being used.
+        std::env::set_var(&var, "111");
+        let result = fetch_svc.template_string(&template, &env_vars_empty);
+        assert!(result.is_ok());
+        assert_eq!(result.unwrap(), "111");
+
+        // When var is present dataset_env_vars, it has top priority.
+        let result = fetch_svc.template_string(&template, &env_vars);
+        assert!(result.is_ok());
+        assert_eq!(result.unwrap(), "222");
+
+        // When dataset_env_vars is empty and system env var is not set,
+        // we expect error.
+        std::env::remove_var(&var);
+        let result = fetch_svc.template_string(&template, &env_vars_empty);
+        assert_matches!(result, Err(Internal(_)));
+
+        // The only supported prefix is "env."
+        let result = fetch_svc.template_string("${{ foo.bar }}", &env_vars_empty);
+        assert!(result.is_err());
+    }
+
+    #[test]
+    fn test_template_string() {
+        let lookup_fn = |env_name: &str| -> Option<String> {
+            match env_name {
+                "test1" => Some(String::from("val1")),
+                "test2" => Some(String::from("val2")),
+                _ => None,
+            }
+        };
+
+        // Populates multiple vars.
+        let result = template_string("${{ env.test1 }}, ${{ env.test2 || default2 }}", &lookup_fn);
+        assert!(result.is_ok());
+        assert_eq!(result.unwrap(), "val1, val2");
+
+        // When at least one of vars is missing, returns error.
+        let result = template_string("{{ env.test1 }}, ${{ env.missing_var }}", &lookup_fn);
+        assert!(result.is_err());
+
+        // Default value is being used if present.
+        let result = template_string("foo=${{ env.test3 || default-val3 }}", &lookup_fn);
```

Comment:
Sorry for not specifying this in the task requirements, but I think we should require quoting the string values. This is to in future allow expressions like: `${{ env.specific || env.general || 'default'}}` where `env.general` is treated as code rather than string value.

The world of templating engines is a complete mess with no standards. I based this super primitive regex implementation on [GitHub Actions](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#using-the-vars-context-to-access-configuration-variable-values) so let's continue with their single-quoted string syntax for now. Later on we should pick a proper templating engine and standardize around it.

Reply:
Will do 👍 As far as I'm touching templates, WDYT about support them in image field es well?
```
image: "ghcr.io/kamu-data/fetch-com.defillama:${{ env.image_ver || '0.1.5' }}"
```

Reply:
Sure, that might be helpful in some cases.